### ISSUE TITLE
Apparent interlacing

### ISSUE THREAD

**Issue by padreputativo** [defold/extension-videoplayer-mpeg#9]
The videos that don’t appear interlaced in other players, or even in the library’s own player, do appear interlaced in Defold. I think it might have to do with the fact that if the texture isn’t perfectly square, the video doesn’t display fully, and if it is, this effect might appear. It’s probably related to printing a texture using UV mapping.

https://github.com/user-attachments/assets/cf5adc01-5941-4429-9224-df736203c99c

**Comment by britzl** [comment]
> I think it might have to do with the fact that if the texture isn’t perfectly square, the video doesn’t display fully, and if it is, this effect might appear. It’s probably related to printing a texture using UV mapping.

Have you tried changing how it is rendering the video texture?

**Comment by padreputativo** [comment]
The file is not interlaced, I’ve checked it with FFmpeg and Blender. According to the library’s author, it doesn’t use interlacing either. It must be that when extracting the texture, the calculation isn’t being done correctly, as you can see in this image with the scale set to 4.
![Screenshot from 2024-09-10 10-47-06](https://github.com/user-attachments/assets/2217f28b-e218-4151-ac4f-16602f474a05)

**Comment by padreputativo** [comment]
> > I think it might have to do with the fact that if the texture isn’t perfectly square, the video doesn’t display fully, and if it is, this effect might appear. It’s probably related to printing a texture using UV mapping.
> 
> Have you tried changing how it is rendering the video texture?

I'm thinking that maybe it's because I had already modified the script before, as it was set to a different resolution than mine and was distorting the image.
![Screenshot from 2024-09-10 10-59-19](https://github.com/user-attachments/assets/34f3f133-afcc-41cd-9b81-7883be694d5e)

**Comment by padreputativo** [comment]
It seems that's not the case; I just tried the version from the repository, and it's doing the same thing.

**Comment by padreputativo** [comment]
![image](https://github.com/user-attachments/assets/b27f4126-5860-4f81-847e-7137fbf2c36f)
It seems that's not the case either; it also appears in the Big Buck Bunny video (though I’m not sure what resolution it has).

**Comment by padreputativo** [comment]
I've tried changing sizes, scale, the atlas, the material... but I suppose that if increasing the scale also increases the size of the interlacing, it has more to do with the frame extraction itself.

**Comment by padreputativo** [comment]
More information from the library creator: https://github.com/phoboslab/pl_mpeg/issues/44

**Comment by padreputativo** [comment]
I tried dumping the frame information to a PNG file with the extension, and it seems to be corrupted.
![test](https://github.com/user-attachments/assets/d28675ba-58ff-4e19-87cf-6b6d06b5efb0)

https://github.com/britzl/defold-png

**Comment by JCash** [comment]
I'd say that since we can produce a texture in the scripts, the data is ok, and you need to look at how you produce the data for the png writer.

**Comment by padreputativo** [comment]
Lua interprets the type as userdata and I'm not sure if it has methods or if it's a table or what is inside it, but converting it to a string to pass it through the extension might result in losing all the information.

![image](https://github.com/user-attachments/assets/23271768-ed54-41d8-854f-eedbf5038fd0)

**Comment by padreputativo** [comment]
Got it:

`local text = buffer.get_bytes(videoframe, "<unknown>")`

![my](https://github.com/user-attachments/assets/bbaa54f2-e0e6-4c6c-af11-2ece2762a59a)

**Comment by padreputativo** [comment]
It seems to be the line that flips the image:

`//int d_index = (rows - row - 1) * 2 * stride;`
`int d_index = row * 2 * stride;`

![image](https://github.com/user-attachments/assets/7e0b0567-e1d8-4fed-8270-dc9f8f687a22)

**Comment by padreputativo** [comment]
I managed to do it, but I'm not sure if you'll like the way I've done it. Since some information is calculated every two rows by default in MPEG, what I did was use the `plm_frame_to_rgb` function from the library itself and then add an additional step to perform a flip with the following code:

`void frame_flip(uint8_t *dest, int width, int height, int stride) {
    int row_size = width * 3;
    uint8_t *temp_row = new uint8_t[row_size];

    for (int row = 0; row < height / 2; row++) {
        uint8_t *top_row = dest + row * stride;
        uint8_t *bottom_row = dest + (height - row - 1) * stride;

        memcpy(temp_row, top_row, row_size);

        memcpy(top_row, bottom_row, row_size);
        memcpy(bottom_row, temp_row, row_size);
    }

    delete[] temp_row;
}`

![image](https://github.com/user-attachments/assets/459cf074-6f42-41f0-b853-5917dc2db159)

**Comment by JCash** [comment]
Nice find.

However, as we're concerned about performance, I do wonder if 1) the decoder has a flip options, or 2), we should simply use a separate shader, letting the GPU flip the uv coordinates.

The GPU is a _lot_ faster than our CPU flip, so it might make sense to use a separate shader instead.

**Comment by padreputativo** [comment]
I totally agree; in fact, I had it implemented in my case with the scale set to -1 on the sprite precisely because I felt it would consume less, as those calculations would still be handled by OpenGL when placing the sprite, even with the scale set to -1. So I had it done that way, but I assumed that perhaps this option was more user-friendly. However, the -1 scale trick is very simple. I'm not sure about that, although in that case, the YUV to RGB conversion and the flip would be done simultaneously, potentially making it even faster than the -1 scale option and CPU conversion.

**Comment by britzl** [comment]
I'm also thinking that it's better to do it in a shader. And like you said @padreputativo I think there was some mention about doing the YUB to RGB conversion using a shader instead. Flipping at the same time would be a nice bonus.

**Comment by padreputativo** [comment]
I've been trying for a while to extract the data directly in YUV, but I haven't been able to do it. I also haven't seen a way to do it without converting to RGB, but it's probably because I'm missing something... If looping is mandatory, I don't think it would make much of a difference whether the calculations are done there or in a shader...

**Comment by JCash** [comment]
Doing the work in a shader is a huge performance win.

**Comment by britzl** [comment]
We have no way for the native extension system to hook into the audio system in Defold. One option is to replace the Defold sound system and use something else. Examples of this:

* https://defold.com/assets/fmod/ (adds an additional sound system)
* https://defold.com/assets/blip/ (removes the Defold sound system and uses openal)

Another option is to extract the audio stream and play that separately as an ogg file.

**Comment by padreputativo** [comment]
I'm not sure if it's supposed to look like that in YUV, and I think it probably shouldn't, haha.

```
void plm_frame_to_yuv(plm_frame_t *frame, uint8_t *dest, int stride) {
    int y_size = frame->width * frame->height;
    int uv_size = (frame->width / 2) * (frame->height / 2);

    memcpy(dest, frame->y.data, y_size);
    memcpy(dest + y_size, frame->cb.data, uv_size);
    memcpy(dest + y_size + uv_size, frame->cr.data, uv_size);
}
```

![image](https://github.com/user-attachments/assets/c0c408c1-7d66-48c3-9b4a-52acddef5ebe)

**Comment by padreputativo** [comment]
To store YUV data in an RGB texture, a prior conversion of the data is necessary. This involves processing all the information through a couple of loops. The way it is done in the official repository is by separating the three layers into different textures (or rather data banks, since the U and V layers contain negative values), which makes it easier to convert them. However, using a single texture and having to convert it beforehand doesn’t seem to make much sense. (At least I’ve learned to create shaders in Defold.) Perhaps it would work with a YCRCB texture format instead of RGB. There is the option to use three textures, but I’m not familiar enough with Defold and/or C++ to know how to do it. It would be easier to read in the shader because it would involve using UVs from 0 to 1 and everything would always be where it should be.

**Comment by britzl** [comment]
I took another look at the YCrCb to RGB conversion and the flip I did was a bit naive, because the decoder is not purely writing pixels one line after the other. There is also a stride and it writes two color values in one row and two in another. When naively flipping the indices the result looks wrong.

This should be the correct conversion and flip if I'm not mistaken. Try this instead and see if you get the results you are after. To my eye it looks correct at least:

```c
void plm_frame_to_rgb_flip(plm_frame_t *frame, uint8_t *dest, int stride) {
    int cols = frame->width >> 1;
    int rows = frame->height >> 1;
    int yw = frame->y.width;
    int cw = frame->cb.width;
    for (int row = 0; row < rows; row++) {
        int c_index = row * cw;
        int y_index = row * 2 * yw;
        int d_index = (rows - row - 1) * 2 * stride;
        for (int col = 0; col < cols; col++) {
            int y;
            int cr = frame->cr.data[c_index] - 128;
            int cb = frame->cb.data[c_index] - 128;
            int r = (cr * 104597) >> 16;
            int g = (cb * 25674 + cr * 53278) >> 16;
            int b = (cb * 132201) >> 16;
            y = ((frame->y.data[y_index + 0]-16) * 76309) >> 16;
            dest[d_index + stride + RI] = plm_clamp(y + r);
            dest[d_index + stride + GI] = plm_clamp(y - g);
            dest[d_index + stride + BI] = plm_clamp(y + b);
            y = ((frame->y.data[y_index + 1]-16) * 76309) >> 16;
            dest[d_index + stride + BYTES_PER_PIXEL + RI] = plm_clamp(y + r);
            dest[d_index + stride + BYTES_PER_PIXEL + GI] = plm_clamp(y - g);
            dest[d_index + stride + BYTES_PER_PIXEL + BI] = plm_clamp(y + b);
            y = ((frame->y.data[y_index + yw]-16) * 76309) >> 16;
            dest[d_index + 0 + RI] = plm_clamp(y + r);
            dest[d_index + 0 + GI] = plm_clamp(y - g);
            dest[d_index + 0 + BI] = plm_clamp(y + b);
            y = ((frame->y.data[y_index + yw + 1]-16) * 76309) >> 16;
            dest[d_index + BYTES_PER_PIXEL + RI] = plm_clamp(y + r);
            dest[d_index + BYTES_PER_PIXEL + GI] = plm_clamp(y - g);
            dest[d_index + BYTES_PER_PIXEL + BI] = plm_clamp(y + b);            c_index += 1;
            y_index += 2;
            d_index += 2 * BYTES_PER_PIXEL;
        }
    }
}
```

**Comment by britzl** [comment]
> Another thing that would be very useful is to be able to use the repository as a dependency since, as there is currently no release, it is not possible.

It is still possible:

https://github.com/defold/extension-videoplayer-mpeg/archive/refs/heads/master.zip

But it is not recommended to depend on the master branch because it can change at any time. Let's make a release after we fix the interlacing!

**Comment by padreputativo** [comment]
Now it looks perfect! (or at least I think so) Yes, the problem was that the 'Y' layer is full resolution, but 'U' and 'V' are at half resolution.

**Comment by britzl** [comment]
Closed via https://github.com/defold/extension-videoplayer-mpeg/commit/59918fa5e030d1381540fd7a49b9b02e753b52c9

**Comment by britzl** [comment]
> Let's make a release after we fix the interlacing!

Released 1.0.0: https://github.com/defold/extension-videoplayer-mpeg/releases/tag/1.0.0

**Comment by padreputativo** [comment]
Many thanks!
